sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,157 @@
+ * JsonFilters assumes that:
+ *   - reset() is called before any skipRow() calls for new row.
+ *   - skipRow() can be called for any valid index of the struct fields,
+ *      and in any order.
+ *   - After skipRow() returns `true`, the internal state of `JsonFilter` can be inconsistent,
+ *     so, skipRow() must not be called for the current row anymore without reset().
+ *
+ * @param pushedFilters The pushed down source filters. The filters should refer to
+ *                      the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(pushedFilters: Seq[sources.Filter], schema: StructType)
+  extends StructFilters(pushedFilters, schema) {
+
+  /**
+   * Stateful JSON predicate that keeps track of its dependent references in the
+   * current row via `refCount`.
+   *
+   * @param predicate The predicate compiled from pushed down source filters.
+   * @param totalRefs The total amount of all filters references which the predicate
+   *                  compiled from.
+   * @param refCount The current number of predicate references in the row that have
+   *                 been not set yet. When `refCount` reaches zero, the predicate
+   *                 has all dependencies are set, and can be applied to the row.
+   */
+  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
+    def reset(): Unit = {
+      refCount = totalRefs
+    }
+  }
+
+  // Predicates compiled from the pushed down filters. The predicates are grouped by their
+  // attributes. The i-th group contains predicates that refer to the i-th field of the given
+  // schema. A predicates can be places to many groups if it has many attributes. For example:
+  //  schema: i INTEGER, s STRING
+  //  filters: IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc"))
+  //  predicates:
+  //    0: Array(IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  //    1: Array(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  private val predicates: Array[Array[JsonPredicate]] = {
+    val groupedPredicates = Array.fill(schema.length)(Array.empty[JsonPredicate])
+    if (SQLConf.get.jsonFilterPushDown) {
+      val groupedByRefSet = filters
+        // Group filters that have the same set of references. For example:
+        //   IsNotNull("i") -> Set("i"), AlwaysTrue -> Set(),
+        //   And(EqualTo("i", 0), StringStartsWith("s", "abc")) -> Set("i", "s")
+        // By grouping filters we could avoid tracking their state of references in the
+        // current row separately.
+        .groupBy(_.references.toSet)
+        // Combine all filters from the same group by `And` because all filters should
+        // return `true` to do not skip a row. The result is compiled to a predicate.
+        .map { case (refSet, refsFilters) =>
+          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
```

Comment:
I thought the initial value of `refCount` should be `refSet.size`, but it's 0?

Reply:
It doesn't matter, actually because `reset()` must be invoke before any `skipRow()` calls. See my comments `JsonFilters assumes that:`